File: benchmark/python/ufunc/arithmetic.py
"""
Universal function benchmarks: arithmetic and math operations
"""

import numpy as np
from ..lib.benchmark_runner import BenchmarkRunner
from ..lib.config import BENCHMARK_CONFIG, ARRAY_SIZES, RANDOM_SEED


def format_size(size: int) -> str:
    """Format a size number for display."""
    if size >= 1_000_000:
        return f"{size // 1_000_000}M"
    elif size >= 1_000:
        return f"{size // 1_000}K"
    return str(size)


UFUNC_MAP = {
    'add': {'fn': np.add, 'binary': True},
    'subtract': {'fn': np.subtract, 'binary': True},
    'multiply': {'fn': np.multiply, 'binary': True},
    'divide': {'fn': np.divide, 'binary': True},
    'sqrt': {'fn': np.sqrt, 'binary': False, 'needs_positive': True},
    'exp': {'fn': np.exp, 'binary': False},
    'log': {'fn': np.log, 'binary': False, 'needs_positive': True},
    'sin': {'fn': np.sin, 'binary': False},
    'cos': {'fn': np.cos, 'binary': False},
    'power': {'fn': np.power, 'binary': True},
}

UFUNC_NAMES = {
    'add': 'Add',
    'subtract': 'Subtract',
    'multiply': 'Multiply',
    'divide': 'Divide',
    'sqrt': 'Square Root',
    'exp': 'Exponential',
    'log': 'Logarithm',
    'sin': 'Sine',
    'cos': 'Cosine',
    'power': 'Power',
}


def benchmark_ufunc(name: str, runner: BenchmarkRunner) -> dict:
    """Benchmark a ufunc across array sizes."""
    results = []
    ufunc = UFUNC_MAP[name]

    for size in ARRAY_SIZES:
        print(f"    {name} size={format_size(size)}... ", end='', flush=True)

        # Generate appropriate data
        np.random.seed(RANDOM_SEED)
        if ufunc.get('needs_positive'):
            arr1 = np.random.uniform(0.1, 1.1, size)
        else:
            arr1 = np.random.randn(size)

        arr2 = None
        if ufunc['binary']:
            np.random.seed(RANDOM_SEED + 1)
            arr2 = np.random.randn(size)

        if ufunc['binary']:
            result = runner.run(lambda: ufunc['fn'](arr1, arr2))
        else:
            result = runner.run(lambda: ufunc['fn'](arr1))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': name, 'name': UFUNC_NAMES[name], 'results': results}


def run_ufunc_benchmarks() -> list[dict]:
    """Run all ufunc benchmarks."""
    runner = BenchmarkRunner(BENCHMARK_CONFIG)
    operations = ['add', 'subtract', 'multiply', 'divide', 'sqrt', 'exp', 'log', 'sin', 'cos', 'power']
    results = []

    for op in operations:
        print(f"  Benchmarking {op}...")
        results.append(benchmark_ufunc(op, runner))

    return results


if __name__ == '__main__':
    import json
    results = run_ufunc_benchmarks()
    print(json.dumps(results, indent=2))

File: packages/numwasm/benchmark/ts/ufunc/arithmetic.ts
/**
 * Universal function benchmarks: arithmetic and math operations
 */

import { BenchmarkRunner } from '../lib/BenchmarkRunner.js';
import { BENCHMARK_CONFIG, ARRAY_SIZES, RANDOM_SEED } from '../lib/config.js';
import { generateRandomData, generateUniformData, formatSize } from '../lib/utils.js';
import type { OperationResults, BenchmarkDataPoint } from '../lib/types.js';

// @ts-expect-error - Types are in dist/index.d.ts but TS can't resolve mjs
import { NDArray, add, subtract, multiply, divide, sqrt, exp, log, sin, cos, power, loadWasmModule } from '../../../numjs.mjs';

type UfuncName = 'add' | 'subtract' | 'multiply' | 'divide' | 'sqrt' | 'exp' | 'log' | 'sin' | 'cos' | 'power';

const ufuncMap: Record<UfuncName, { fn: Function; binary: boolean; needsPositive?: boolean }> = {
  add: { fn: add, binary: true },
  subtract: { fn: subtract, binary: true },
  multiply: { fn: multiply, binary: true },
  divide: { fn: divide, binary: true },
  sqrt: { fn: sqrt, binary: false, needsPositive: true },
  exp: { fn: exp, binary: false },
  log: { fn: log, binary: false, needsPositive: true },
  sin: { fn: sin, binary: false },
  cos: { fn: cos, binary: false },
  power: { fn: power, binary: true },
};

const ufuncNames: Record<UfuncName, string> = {
  add: 'Add',
  subtract: 'Subtract',
  multiply: 'Multiply',
  divide: 'Divide',
  sqrt: 'Square Root',
  exp: 'Exponential',
  log: 'Logarithm',
  sin: 'Sine',
  cos: 'Cosine',
  power: 'Power',
};

async function benchmarkUfunc(
  name: UfuncName,
  runner: BenchmarkRunner
): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];
  const ufunc = ufuncMap[name];

  for (const size of ARRAY_SIZES) {
    process.stdout.write(`    ${name} size=${formatSize(size)}... `);

    // Generate appropriate data
    let data1: number[];
    if (ufunc.needsPositive) {
      // Use uniform [0.1, 1.1] for positive values (avoid 0 for log/sqrt)
      data1 = generateUniformData(size, RANDOM_SEED).map(x => x + 0.1);
    } else {
      data1 = generateRandomData(size, RANDOM_SEED);
    }

    const arr1 = await NDArray.fromArray(data1);
    let arr2: typeof arr1 | null = null;

    if (ufunc.binary) {
      const data2 = generateRandomData(size, RANDOM_SEED + 1);
      arr2 = await NDArray.fromArray(data2);
    }

    try {
      const result = await runner.run(
        () => ({ arr1, arr2 }),
        (setup) => {
          let res: typeof arr1;
          if (ufunc.binary && setup.arr2) {
            res = ufunc.fn(setup.arr1, setup.arr2);
          } else {
            res = ufunc.fn(setup.arr1);
          }
          const val = res.sum();
          res.dispose();
          return val;
        }
      );

      results.push({
        params: { size },
        meanMs: result.meanMs,
        minMs: result.minMs,
        maxMs: result.maxMs,
        stdMs: result.stdMs,
        iterations: result.iterations,
        result: result.result,
      });

      console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
    } finally {
      arr1.dispose();
      if (arr2) arr2.dispose();
    }
  }

  return { id: name, name: ufuncNames[name], results };
}

export async function runUfuncBenchmarks(): Promise<OperationResults[]> {
  console.log('  Loading WASM module...');
  await loadWasmModule();

  const runner = new BenchmarkRunner(BENCHMARK_CONFIG);
  const operations: UfuncName[] = ['add', 'subtract', 'multiply', 'divide', 'sqrt', 'exp', 'log', 'sin', 'cos', 'power'];
  const results: OperationResults[] = [];

  for (const op of operations) {
    console.log(`  Benchmarking ${op}...`);
    results.push(await benchmarkUfunc(op, runner));
  }

  return results;
}
